cpp/dolfinx/fem/DirichletBC.h

```diff
@@ -495,76 +508,82 @@ class DirichletBC
   void set(std::span<T> x, std::optional<std::span<const T>> x0,
            T alpha = 1) const
   {
-    std::int32_t x_size = x.size();
-    if (alpha == T(0)) // Optimisation for when alpha == 0
+    // set_fn is a lambda which gets evaluated for every index in [0,
+    // _dofs0.size()) and its result is assigned to x[_dofs0[i]].
+    auto apply = [&](std::invocable<std::int32_t> auto set_fn)
     {
-      for (std::int32_t idx : _dofs0)
+      static_assert(
+          std::is_same_v<std::invoke_result_t<decltype(set_fn), std::int32_t>,
+                         T>);
+
+      std::int32_t x_size = x.size();
+      for (std::int32_t i = 0; i < _dofs0.size(); ++i)
       {
-        if (idx < x_size)
-          x[idx] = 0;
+        if (_dofs0[i] < x_size)
+          x[_dofs0[i]] = set_fn(i);
       }
+    };
+
+    if (alpha == T(0)) // Optimisation for when alpha == 0
+    {
+      apply([](std::int32_t) -> T { return 0; });
+      return;
     }
-    else
+
+    auto handle_function = [&](std::shared_ptr<const Function<T, U>> g)
```

Comment:
Lambda function name is cryptic. `function` is a highly overloaded word, not sure what `handle` is.

Function should take a shared pointer - it doesn't hold onto it. Pass a reference.

Reply:
We cannot change the type to just a function here, as we use it for the dispatch of the `std::variant`. Depending on which value it holds the "handle" for the specified type is called by `std::visit`. "Function" is chosen not to show that we are defining a function here, but refers to the type this lambda handles.

Reply:
We tend to use `fn` rather than `function` elsewhere to denote a C++ 'function' rather than a DOLFINx 'Function'.